// react-search-bar/lib/suggestions.js
'use strict';

Object.defineProperty(exports, '__esModule', {
  value: true
});

var _createClass = (function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ('value' in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; })();

var _get = function get(_x, _x2, _x3) { var _again = true; _function: while (_again) { var object = _x, property = _x2, receiver = _x3; _again = false; if (object === null) object = Function.prototype; var desc = Object.getOwnPropertyDescriptor(object, property); if (desc === undefined) { var parent = Object.getPrototypeOf(object); if (parent === null) { return undefined; } else { _x = parent; _x2 = property; _x3 = receiver; _again = true; desc = parent = undefined; continue _function; } } else if ('value' in desc) { return desc.value; } else { var getter = desc.get; if (getter === undefined) { return undefined; } return getter.call(receiver); } } };

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { 'default': obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError('Cannot call a class as a function'); } }

function _inherits(subClass, superClass) { if (typeof superClass !== 'function' && superClass !== null) { throw new TypeError('Super expression must either be null or a function, not ' + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

var _reactAutobind = require('react-autobind');

var _reactAutobind2 = _interopRequireDefault(_reactAutobind);

var _classnames = require('classnames');

var _classnames2 = _interopRequireDefault(_classnames);

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var Suggestions = (function (_React$Component) {
  _inherits(Suggestions, _React$Component);

  function Suggestions(props) {
    _classCallCheck(this, Suggestions);

    _get(Object.getPrototypeOf(Suggestions.prototype), 'constructor', this).call(this, props);
    this.state = {
      activeItem: -1
    };
    (0, _reactAutobind2['default'])(this);
  }

  _createClass(Suggestions, [{
    key: 'onMouseEnter',
    value: function onMouseEnter(index) {
      this.setState({
        activeItem: index
      });
    }
  }, {
    key: 'onMouseLeave',
    value: function onMouseLeave() {
      this.setState({
        activeItem: -1
      });
    }
  }, {
    key: 'onMouseDown',
    value: function onMouseDown(event) {
      event.preventDefault();
    }
  }, {
    key: 'onTouchStart',
    value: function onTouchStart(activeItem) {
      var _this = this;

      this.timer = setTimeout(function () {
        return _this.setState({
          activeItem: activeItem
        });
      }, 200);
    }
  }, {
    key: 'onTouchMove',
    value: function onTouchMove() {
      clearTimeout(this.timer);
      this.touchedMoved = true;
      this.setState({ activeItem: -1 });
    }
  }, {
    key: 'onTouchEnd',
    value: function onTouchEnd(suggestion) {
      var _this2 = this;

      if (!this.touchedMoved) {
        setTimeout(function () {
          return _this2.props.onSelection(suggestion);
        }, 220);
      }
      this.touchedMoved = false;
    }
  }, {
    key: 'render',
    value: function render() {
      var _this3 = this;

      var _props = this.props;
      var highlightedItem = _props.highlightedItem;
      var searchTerm = _props.searchTerm;
      var suggestions = _props.suggestions;
      var activeItem = this.state.activeItem;

      return _react2['default'].createElement(
        'ul',
        {
          className: 'search-bar-suggestions',
          onMouseLeave: this.onMouseLeave
        },
        suggestions.map(function (suggestion, index) {
          return _react2['default'].createElement(
            'li',
            {
              className: (0, _classnames2['default'])({
                highlighted: highlightedItem === index || activeItem === index
              }),
              key: index,
              onClick: _this3.props.onSelection.bind(_this3, suggestion),
              onMouseEnter: _this3.onMouseEnter.bind(_this3, index),
              onMouseDown: _this3.onMouseDown,
              onTouchStart: _this3.onTouchStart.bind(_this3, index),
              onTouchMove: _this3.onTouchMove,
              onTouchEnd: _this3.onTouchEnd
            },
            _react2['default'].createElement(
              'span',
              null,
              searchTerm,
              _react2['default'].createElement(
                'strong',
                null,
                suggestion.substr(searchTerm.length)
              )
            )
          );
        })
      );
    }
  }]);

  return Suggestions;
})(_react2['default'].Component);

Suggestions.propTypes = {
  highlightedItem: _react.PropTypes.number,
  searchTerm: _react.PropTypes.string.isRequired,
  suggestions: _react.PropTypes.array.isRequired
};

exports['default'] = Suggestions;
module.exports = exports['default'];
